src/core/Refiner.ml

```diff
@@ -891,6 +842,75 @@ struct
     S.Com (fam, src, trg, cof, tm), vfam_trg
 end
 
+
+module Signature =
+struct
+  let formation (tacs : T.Tp.tac telescope) : T.Tp.tac =
+    let rec form_fields tele =
+      function
+      | Bind (lbl, tac, tacs) ->
+        let* tp = T.Tp.run tac in
+        let* vtp = RM.lift_ev @@ Sem.eval_tp tp in
+        T.abstract ~ident:(lbl :> Ident.t) vtp @@ fun var -> form_fields (Snoc (tele, (lbl, tp))) (tacs var)
+      | Done -> RM.ret @@ S.Signature (Bwd.to_list tele)
+    in T.Tp.rule ~name:"Signature.formation" @@ form_fields Emp tacs
+
+  let rec find_field_tac (fields : (Ident.user * T.Chk.tac) list) (lbl : Ident.user) : T.Chk.tac option =
+    match fields with
+    | (lbl', tac) :: _ when Ident.equal lbl lbl'  ->
+      Some tac
+    | _ :: fields ->
+      find_field_tac fields lbl
+    | [] ->
+      None
+
+
+  let rec intro_fields phi phi_clo (sign : D.sign) (tacs : Ident.user -> T.Chk.tac option) : (Ident.user * S.t) list m =
+    match sign with
+    | D.Field (lbl, tp, sign_clo) ->
+      let tac = match tacs lbl, tp with
+        | Some tac, _ -> tac
+        | None, ElStable (`Ext (0,_ ,`Global (Cof (Meet [])), _)) -> Univ.infer_nullary_ext
```

Comment:
I suggest here rather than matching on `Meet []` that you use the monad to check whether the cofibiration is true (i.e. equivalent to `Meet []`). Otherwise this refinement rule will not be stable under definitional equivalence. Does that make sense?

Reply:
Yeah, that does make sense. Because of equality reflection for cofibrations, this would treat cofibrations that are, as you say, definitionally equal to `Meet []`, as being different from `Meet []`, right?

Reply:
Yes, that's right!